wpilibc/src/main/native/include/frc/trajectory/constraints/TrajectoryConstraint.h

```diff
@@ -0,0 +1,65 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <limits>
+
+#include <units/units.h>
+
+#include "frc/geometry/Pose2d.h"
+#include "frc/spline/Spline.h"
+
+namespace frc {
+/**
+ * An interface for defining user-defined velocity and acceleration constraints
+ * while generating trajectories.
+ */
+class TrajectoryConstraint {
+ public:
+  /**
+   * Represents a minimum and maximum acceleration.
+   */
+  struct MinMax {
+    /**
+     * The minimum acceleration.
+     */
+    units::meters_per_second_squared_t minAcceleration{
+        -std::numeric_limits<double>::max()};
+
+    /**
+     * The maximum acceleration.
+     */
+    units::meters_per_second_squared_t maxAcceleration{
+        std::numeric_limits<double>::max()};
+  };
+
+  /**
+   * Returns the max velocity given the current pose and curvature.
+   *
+   * @param pose The pose at the current point in the trajectory.
+   * @param curvature The curvature at the current point in the trajectory.
+   *
+   * @return The absolute maximum velocity.
+   */
+  virtual units::meters_per_second_t MaxVelocity(const Pose2d& pose,
+                                                 curvature_t curvature) = 0;
+
+  /**
+   * Returns the minimum and maximum allowable acceleration for the trajectory
+   * given pose, curvature, and speed.
+   *
+   * @param pose The pose at the current point in the trajectory.
+   * @param curvature The curvature at the current point in the trajectory.
+   * @param speed The speed at the current point in the trajectory.
+   *
+   * @return The min and max acceleration bounds.
+   */
+  virtual MinMax MinMaxAcceleration(const Pose2d& pose, curvature_t curvature,
```

Comment:
What's the reasoning for making these pure virtual? In what ways will the user reasonably customize what it returns? Do we intend to provide canned constraint classes like the centripetal acceleration one in the tests? Otherwise, I don't see a point to putting just this class in a "constraint" folder.

Reply:
Yeah, my intention was to provide some canned constraints (with the centripetal acceleration constraint being one of them). Some other examples of user-defined constraints could include restricting the robot's velocity or acceleration in a specific region (useful for the 2019 HAB platform) or introducing a constraint that uses the differential drive kinematics to ensure that both wheel speeds are under some specified max velocity.

Reply:
I think the diff drive one will help us home in on the desired API. I suspect I'll be doing some function/struct name bikeshedding in this class later.